docs(affinity): updates the pod scheduling section to use node pools

PaulRMellor opened the pull request:
**Documentation**

- Updates pod scheduling concepts and procedures
- Procedures now use node pools as examples
- Concepts streamlined and made more concise
- Kafka config option dropped and cruise control config added to concepts
- Minor restructure of storage section to make storage affinity and metadata storage more visible in the ToC 

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

scholzj commented on documentation/modules/configuring/proc-dedicated-nodes.adoc:
```diff
@@ -3,61 +3,78 @@
 
 [id='proc-dedicated-nodes-{context}']
-= Setting up dedicated nodes and scheduling pods on them
+= Configuring pod affinity for dedicated nodes
```
This is way more than affinity. So I think we should keepo the oridinal title.

scholzj commented on documentation/modules/configuring/proc-dedicated-nodes.adoc:
I can probably live with this. But I think there are some things to consider:
* The old content was technically correct, and I think it was the more common way to do it
* When/why do we expect users to configure dedicated nodes only for one node pool instead of the whole cluster? In general, we assume that production clusters (which is where dedicated nodes would be typically used) have at least two node pools.
* Should we tell users that this can be configured through the Kafka CR as well?
* Should we warn users about the risk of duplicate fields in Kafka and KafkaNodePool CR? (I think the other points are up for discussion, but here the clear answer is IMHO that we should mention it)

scholzj commented on documentation/modules/configuring/proc-scheduling-based-on-other-pods.adoc:
I guess the previous comments apply here as well.
* We should probably warn about the duplicate template beween KAfka and KafkaNodePool
* Do we need to make it more clear why would you configure this per-node pool?

scholzj commented on documentation/modules/configuring/proc-scheduling-brokers-on-different-worker-nodes.adoc:
I think here the use of Kafka node pools might be controversial again. Are you fine if your brokers share node with controllers? Are you fine if they share the node with brokers from other node pools or other clusters?

scholzj commented on documentation/modules/configuring/proc-scheduling-deployment-to-node-using-node-affinity.adoc:
I guess this is the chapter where doing it based on KafkaNOdePool makes most sense. If you would want a good example, typically users ask about this to schedule one node pool on AZ1, second node pool on AZ2 etc. That is probably more common in this scenario than using the fast network.

scholzj reviewed (commented):
I left some comments. But TBH, I'm strugling to see why we are doing this and I'm not convined this is the right way.

I think the old docs were fully functional. But they show only one of the way how to do it (through the Kafka CR). In most cases, I think it would be the more common way. But it is definitely not the only way. Now we seem to be changing it completely to the other way (KafkaNodePool). But once again, we show only one way how to do it.

I think we need to find a bit more generic way how to present this to make it more clear that you can configure the affinity & tolerations in different ways and that they are both suitable in slightly different situations. This is also important as we have to explain how the two ways interact as they essentially esclude each other.

PaulRMellor commented:
> I left some comments. But TBH, I'm strugling to see why we are doing this and I'm not convined this is the right way.
> 
> I think the old docs were fully functional. But they show only one of the way how to do it (through the Kafka CR). In most cases, I think it would be the more common way. But it is definitely not the only way. Now we seem to be changing it completely to the other way (KafkaNodePool). But once again, we show only one way how to do it.
> 
> I think we need to find a bit more generic way how to present this to make it more clear that you can configure the affinity & tolerations in different ways and that they are both suitable in slightly different situations. This is also important as we have to explain how the two ways interact as they essentially esclude each other.

Thanks @scholzj -- discussing both ways and being generic where possible seems like a sensible approach.

PaulRMellor commented:
For the second draft (second commit), I tried a few ways to present Kafka and node pools config. In the end, I thought it clearest just to show examples for both.  We already have procedure to show [Managing storage affinity using node pools](https://strimzi.io/docs/operators/in-development/deploying#proc-managing-storage-affinity-node-pools-str) with availability zones in the section on storage, so I linked to that from the procedure for Configuring pod scheduling for specific nodes.

scholzj commented on documentation/modules/configuring/proc-dedicated-nodes.adoc:
```diff
@@ -3,49 +3,82 @@
+For more granular control, if needed, you can override the default policy for specific node pools using the `KafkaNodePool` resource.
+This procedure shows example configurations for both.
+
+IMPORTANT: A `nodeAffinity` rule defined in a `KafkaNodePool` resource overrides any rule in the `Kafka` resource. 
```
This is correct, but the situation is a bit more complicated ...
* If there is anything in `spec.template` in KafkaNodePool, the corresponding things in .spec.kafka.template` in Kafka are ignored.
* So, not only if there is node affinity configured in the node pool the node affinity from the Kafka CR will be ignored
* Even for examle if there is a container variable defined in `.spec.template.kafkaContainer` in KafkaNodePool, the affinity form `.spec.kafka.template in Kafka CR will be ignored.

I think we should make this more clear here. Maybe have something in one place explaining it in detail and having a link there from all places like this.

scholzj commented on documentation/modules/configuring/proc-scheduling-based-on-other-pods.adoc:
```diff
@@ -6,26 +6,63 @@
+For more granular control, if needed, you can override the default policy for specific node pools using the `KafkaNodePool` resource.
+This procedure shows example configurations for both.
+
+IMPORTANT: A `podAntiAffinity` rule defined in a `KafkaNodePool` resource overrides any rule in the `Kafka` resource. 
```
Same as above.

scholzj commented on documentation/modules/configuring/proc-scheduling-brokers-on-different-worker-nodes.adoc:
```diff
@@ -6,29 +6,64 @@
+For more granular control, if needed, you can override the default policy for specific node pools using the `KafkaNodePool` resource.
+This procedure shows example configurations for both.
+
+IMPORTANT: A `podAntiAffinity` rule defined in a `KafkaNodePool` resource overrides any rule in the `Kafka` resource. 
```
Same as above.

scholzj commented on documentation/modules/configuring/ref-affinity.adoc:
Slightly unrelated to the core purpose of the PR. But I wonder if we should mention topology spread constraints in here as well. They are another way how to distribute the Pod in order to make them equally distributed accross some _partition_ such as availability zone. (https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/)

scholzj reviewed (commented):
I think the second commit looks good. I left some comments, but I think the structure works for me.

PaulRMellor commented:
Thanks for the review again @scholzj 
I've updated the strategies (concepts) section to include a clarification on how `pod.template` config is handled in Kafka and Node pools. I've also added something on  topology spread constraints

ppatierno commented on documentation/modules/configuring/ref-affinity.adoc:
```diff
@@ -3,59 +3,68 @@
-* Node affinity
+Use node affinity to schedule workloads onto specific nodes::
+Kubernetes clusters often include nodes optimized for different workloads, such as CPU, memory, storage, or network.  
+Node affinity enables scheduling Kafka components onto nodes that match specific labels, such as `beta.kubernetes.io/instance-type` or custom labels, to optimize performance and cost.
```
The `beta.kubernetes.io/instance-type` seems to be deprecated since Kubernetes 1.17, it's replaced by `node.kubernetes.io/instance-type`.
